File: android/app/src/main/java/info/bagen/rust/plaoc/microService/helper/AdapterManager.kt
package info.bagen.rust.plaoc.microService.helper

open class AdapterManager<T> {
    private val adapterOrderMap = mutableMapOf<T, Int>()
    private var orderdAdapters = listOf<T>()
    val adapters get() = orderdAdapters
    fun append(order: Int = 0, adapter: T): (Unit) -> Boolean {
        adapterOrderMap[adapter] = order
        orderdAdapters =
            adapterOrderMap.toList().sortedBy { (_, b) -> b }.map { (adapter) -> adapter }
        return { remove(adapter) }
    }

    fun remove(adapter: T) = adapterOrderMap.remove(adapter) != null
}

File: ios/BFExplorer/Plaoc/IPCCore/helper/AdapterManager.swift
//
//  AdapterManager.swift
//  BFExplorer
//
//  Created by ui03 on 2023/3/15.
//

import Foundation

class AdapterManager<T> {
    
    private var adapterOrderMap = NSMutableDictionary()
    private var orderdAdapters: [T] = []
    
    var adapters: [T] {
        return orderdAdapters
    }
    
    func append(order: Int = 0, adapter: T) -> () -> Bool {
        
        adapterOrderMap[adapter] = order
        orderdAdapters = adapterOrderMap.sorted(by: {($0.1 as! Int) < ($1.1 as! Int)}).map { $0.0 as! T }
        return { self.remove(adapter: adapter) }
    }
    
    func remove(adapter: T) -> Bool {

        if adapterOrderMap.object(forKey: adapter) != nil {
            adapterOrderMap.removeObject(forKey: adapter)
            return true
        }
        return false
    }
}
